onnxruntime/core/providers/cuda/tensor/gather_elements.cc

```diff
@@ -1,101 +1,216 @@
+
+namespace {
+bool CanSkip(const TensorShapeVector& input_shape, const TensorShapeVector& indices_shape, size_t dim) {
+  return input_shape[dim] == 1 && indices_shape[dim] == 1;
+}
+
+bool CanMerge(const TensorShapeVector& input_shape, const TensorShapeVector& indices_shape, size_t src, size_t dst) {
+  return input_shape[src] == indices_shape[src] && input_shape[dst] == indices_shape[dst];
+}
+
+void Move(TensorShapeVector& input_shape, TensorShapeVector& indices_shape, size_t src, size_t dst) {
+  input_shape[dst] = input_shape[src];
+  indices_shape[dst] = indices_shape[src];
+}
+
+void Merge(TensorShapeVector& input_shape, TensorShapeVector& indices_shape, size_t src, size_t dst) {
+  input_shape[dst] *= input_shape[src];
+  indices_shape[dst] *= indices_shape[src];
+}
+}  // namespace
+
+void CoalesceDimensions(TensorShapeVector& input_shape, TensorShapeVector& indices_shape, int64_t axis,
+                        int64_t& new_axis, int64_t& new_rank, int64_t& input_stride_along_axis,
+                        TArray<int64_t>& masked_input_strides, TArray<fast_divmod>& indices_fdms) {
+  size_t rank = input_shape.size();
+  // Reverse for better calculation.
+  std::reverse(input_shape.begin(), input_shape.end());
+  std::reverse(indices_shape.begin(), indices_shape.end());
+  size_t reverse_axis = rank - 1 - static_cast<size_t>(axis);
+  size_t curr = 0, next = 0;
+  while (curr < reverse_axis && CanSkip(input_shape, indices_shape, curr)) ++curr;
+  if (curr < reverse_axis) {
+    if (curr > 0) Move(input_shape, indices_shape, curr, 0);
+    next = curr + 1;
+    curr = 0;
+    while (next < reverse_axis) {
+      if (!CanSkip(input_shape, indices_shape, next)) {
+        if (CanMerge(input_shape, indices_shape, next, curr)) {
+          Merge(input_shape, indices_shape, next, curr);
+        } else {
+          ++curr;
+          if (curr != next) Move(input_shape, indices_shape, next, curr);
+        }
+      }
+      ++next;
+    }
+  }
+
+  if (curr == reverse_axis) {
+    if (curr > 0) Move(input_shape, indices_shape, curr, 0);
+    curr = 0;
+  } else {
+    ++curr;
+    // next is now the reverse_axis
+    if (curr != next) Move(input_shape, indices_shape, next, curr);
+  }
+  size_t new_reverse_axis = curr;
+  next = reverse_axis + 1;
+  while (next < rank && CanSkip(input_shape, indices_shape, next)) ++next;
+  if (next < rank) {
+    ++curr;
+    if (curr != next) Move(input_shape, indices_shape, next, curr);
+    ++next;
+    while (next < rank) {
+      if (!CanSkip(input_shape, indices_shape, next)) {
+        if (CanMerge(input_shape, indices_shape, next, curr)) {
+          Merge(input_shape, indices_shape, next, curr);
+        } else {
+          ++curr;
+          if (curr != next) Move(input_shape, indices_shape, next, curr);
+        }
+      }
+      ++next;
+    }
+  }
+
+  new_rank = curr + 1;
+  new_axis = static_cast<int64_t>(new_rank - 1 - new_reverse_axis);
+  input_shape.resize(new_rank);
```

Comment:
I am wondering is there any reason we did not create a new shape, operate on that, as a returned value in param list?

Reply:
Both way are OK. Using new shapes will need more variables from outside and passed as reference. More variables means more confusion. If the callers don't want to change the old shape instances, they can duplicate them before calling this.